codec/src/types/primitives.rs

```diff
@@ -67,102 +62,154 @@ impl Codec for bool {
         varint::size(len) + self.len()
     }
+}
 
+impl<R: RangeBounds<usize>> Read<R> for Bytes {
     #[inline]
-    fn read(buf: &mut impl Buf) -> Result<Self, Error> {
+    fn read_cfg(buf: &mut impl Buf, cfg: R) -> Result<Self, Error> {
         let len32 = varint::read::<u32>(buf)?;
         let len = usize::try_from(len32).map_err(|_| Error::InvalidVarint)?;
+        if !cfg.contains(&len) {
+            return Err(Error::InvalidLength(len));
+        }
         at_least(buf, len)?;
         Ok(buf.copy_to_bytes(len))
     }
 }
 
 // Constant-size array implementation
-impl<const N: usize> Codec for [u8; N] {
+impl<const N: usize> Write for [u8; N] {
     #[inline]
     fn write(&self, buf: &mut impl BufMut) {
         buf.put(&self[..]);
     }
+}
 
+impl<const N: usize> Read for [u8; N] {
     #[inline]
-    fn len_encoded(&self) -> usize {
-        N
-    }
-
-    #[inline]
-    fn read(buf: &mut impl Buf) -> Result<Self, Error> {
+    fn read_cfg(buf: &mut impl Buf, _: ()) -> Result<Self, Error> {
         at_least(buf, N)?;
         let mut dst = [0; N];
         buf.copy_to_slice(&mut dst);
         Ok(dst)
     }
 }
 
-impl<const N: usize> SizedCodec for [u8; N] {
+impl<const N: usize> FixedSize for [u8; N] {
     const LEN_ENCODED: usize = N;
 }
 
 // Option implementation
-impl<T: Codec> Codec for Option<T> {
+impl<T: Encode> Write for Option<T> {
     #[inline]
     fn write(&self, buf: &mut impl BufMut) {
         self.is_some().write(buf);
         if let Some(inner) = self {
             inner.write(buf);
         }
     }
+}
 
+impl<T: Encode> Encode for Option<T> {
     #[inline]
     fn len_encoded(&self) -> usize {
         match self {
             Some(inner) => 1 + inner.len_encoded(),
             None => 1,
         }
     }
+}
 
+impl<Cfg, T: Read<Cfg>> Read<Cfg> for Option<T> {
     #[inline]
-    fn read(buf: &mut impl Buf) -> Result<Self, Error> {
+    fn read_cfg(buf: &mut impl Buf, cfg: Cfg) -> Result<Self, Error> {
         if bool::read(buf)? {
-            Ok(Some(T::read(buf)?))
+            Ok(Some(T::read_cfg(buf, cfg)?))
         } else {
             Ok(None)
         }
     }
 }
 
+// Vec implementation
+impl<T: Write> Write for Vec<T> {
+    #[inline]
+    fn write(&self, buf: &mut impl BufMut) {
+        let len = u32::try_from(self.len()).expect("Vec length exceeds u32");
+        varint::write(len, buf);
+        for item in self {
+            item.write(buf);
+        }
+    }
+}
+
+impl<T: Encode> Encode for Vec<T> {
+    #[inline]
+    fn len_encoded(&self) -> usize {
+        let len = u32::try_from(self.len()).expect("Vec length exceeds u32");
+        varint::size(len) + self.iter().map(Encode::len_encoded).sum::<usize>()
+    }
+}
+
+impl<R: RangeBounds<usize>, Cfg: Copy, T: Read<Cfg>> Read<(R, Cfg)> for Vec<T> {
+    #[inline]
+    fn read_cfg(buf: &mut impl Buf, (range, cfg): (R, Cfg)) -> Result<Self, Error> {
```

Comment:
I would expect the `cfg` to just be RangeBounds? Is there a reason we need to keep both?

Ahhh for the inner type config.

Reply:
I wonder if it makes sense to implement something like this for just RangeBounds for inner types that have no Cfg? I expect that to be a very common call.